File: server/src/server/video_composition/timeline.py
# 由业务快照生成 IMS Timeline；模板时间规则按成片时长计算，字幕保留文案来源。

from math import floor

from pydantic import TypeAdapter

from ..template.schema import CATEGORY_PARAMETERS, EffectTemplateEditor, Template
from ..template.timing import resolve_track
from .schema import CompositionRequest, MatchedSegment, Segment


def validate_segments(segments: list[Segment], duration_ms: int) -> None:
    """校验非空、唯一编号、单调秒制区间，尾部静音仍由音频总长决定。"""
    if not segments or type(duration_ms) is not int or duration_ms <= 0:
        raise ValueError("缺少有效切片或音频总时长")
    previous = 0
    ids = set()
    for item in segments:
        if item.segment_id in ids or not previous <= item.start_time < item.end_time <= duration_ms / 1000:
            raise ValueError("切片编号重复、时间重叠或超出音频范围")
        ids.add(item.segment_id)
        previous = item.end_time


def validate_matches(segments: list[Segment], matches: list[MatchedSegment]) -> None:
    """逐项核对上游整数编号、顺序、文案和秒制边界，拒绝其他任务或重新对齐的结果。"""
    if len(segments) != len(matches):
        raise ValueError("匹配片段数量不一致")
    for source, matched in zip(segments, matches):
        if (
            matched.segment_id != source.segment_id or matched.text != source.text
            or matched.start_time != source.start_time
            or matched.end_time != source.end_time
        ):
            raise ValueError("匹配片段编号、文案或时间与切片不一致")


def build_timeline(
    request: dict, template: dict, segments: list[dict], matches: list[dict],
    duration_ms: int, *, width: int, height: int, fps: int,
) -> tuple[dict, list[str]]:
    """返回时间线及转场调整说明；无网络、模型、随机值或对输入快照的修改。"""
    request = CompositionRequest.model_validate(request)
    template = Template.model_validate(template)
    source = TypeAdapter(list[Segment]).validate_python(segments)
    matched = TypeAdapter(list[MatchedSegment]).validate_python(matches)
    validate_segments(source, duration_ms)
    validate_matches(source, matched)
    if not all(type(v) is int and v > 0 for v in (width, height, fps)) or fps > 60:
        raise ValueError("输出尺寸或帧率无效")
    duration = duration_ms / 1000
    editor = template.editor
    by_id = {item.id: item for item in template.effects}
    if len(by_id) != len(template.effects):
        raise ValueError("模板效果快照包含重复引用")

    def parameters_for(config: EffectTemplateEditor) -> dict:
        """逐个核对对象的可信效果快照，即使当前没有可显示区间也检查引用。"""
        parameters = {}
        for key, effect_id in config.selected_effects().items():
            category = {
                "title_flower": "flower", "subtitle_flower": "flower", "bubble": "bubble",
                "filter": "filter", "vfx": "vfx/normal", "transition": "transition/normal",
            }.get(key, key.rsplit("_", 1)[-1])
            item = by_id.get(effect_id)
            if (
                item is None or item.category != category or item.id not in template.effect_ids
                or item.parameters != {CATEGORY_PARAMETERS[category]: item.effect_id}
                or item.effect_id == "random"
            ):
                raise ValueError("模板效果引用、类别或参数无效")
            parameters[key] = item.parameters
        return parameters

    parameters = parameters_for(editor) if template.tracks is None else {}

    def video(url: str, kind: str, start: float, end: float, source_start: float) -> dict:
        """数字人源时间等于成片时间；素材从源零点起，所有视频显式静音。"""
        clip = {
            "Type": "Image" if kind == "image" else "Video", "MediaURL": url,
            "TimelineIn": start, "TimelineOut": end,
            "Width": width, "Height": height, "AdaptMode": "Cover",
            "Effects": [{"Type": "Volume", "Gain": 0}] if kind == "video" else [],
        }
        if kind == "image":
            clip["Duration"] = end - start
        else:
            clip.update(In=source_start, Out=end if source_start == start else source_start + (end - start))
        return clip

    # 只按命中素材切开可见画面；连续未命中字幕和所有间隙合为原时刻数字人。
    clips, cursor = [], 0
    for item, match in zip(source, matched):
        if match.matched_candidate_url is None:
            continue
        if cursor < item.start_time:
            clips.append(video(request.video_url, "video", cursor, item.start_time, cursor))
        clips.append(video(match.matched_candidate_url, match.matched_candidate_type,
                           item.start_time, item.end_time, 0))
        cursor = item.end_time
    if cursor < duration:
        clips.append(video(request.video_url, "video", cursor, duration, cursor))

    warnings = []
    transition = parameters.get("transition")
    if transition and len(clips) > 1:
        # 每个片段为相邻两侧均分预算；向下取整到输出帧，短空隙保留直接切换。
        capacities = [
            (clip["TimelineOut"] - clip["TimelineIn"]) / (1 if i in (0, len(clips) - 1) else 2)
            for i, clip in enumerate(clips)
        ]
        for i in range(len(clips) - 1):
            desired = template.transition_duration_seconds
            seconds = floor(min(desired, capacities[i], capacities[i + 1]) * fps + 1e-9) / fps
            seconds = floor(seconds * 10000 + 1e-9) / 10000
            if seconds < desired:
                warnings.append(f"画面边界 {clips[i]['TimelineOut']}s 转场缩短为 {seconds}s")
            if seconds > 0:
                clips[i]["Effects"].append({"Type": "DLTransition", **transition, "Duration": seconds})

    def text(role: str, content: str, start: float, end: float,
             config: EffectTemplateEditor = editor, effects: dict = parameters) -> dict:
        """使用业务文字和模板样式，短入出动画同比缩短且不跨字幕区间。"""
        clip = {
            "Type": "Text", "Content": content, "TimelineIn": start,
            "TimelineOut": end, "Alignment": "Center",
            "X": min(getattr(config, f"{role}_x") / 100, 0.9999),
            "Y": min(getattr(config, f"{role}_y") / 100, 0.9999),
            "Font": "Alibaba PuHuiTi", "FontSize": getattr(config, f"{role}_size"),
            "FontColor": "#FFFFFF", "Outline": 0, "AdaptMode": "AutoWrap",
            **effects.get("bubble" if role == "bubble" else f"{role}_flower", {}),
        }
        timings = {motion: getattr(config, f"{role}_{motion}_duration")
                   for motion in ("in", "out") if f"{role}_{motion}" in effects}
        total = sum(timings.values())
        scale = min(1, (end - start) / total) if total else 1
        for motion in ("in", "out", "loop"):
            clip.update(effects.get(f"{role}_{motion}", {}))
            if motion in timings:
                seconds = floor(timings[motion] * scale * 10000 + 1e-9) / 10000
                if seconds <= 0:
                    raise ValueError("字幕太短，无法表达所选动画")
                clip[f"AaiMotion{motion.title()}"] = seconds
        return clip

    subtitle_tracks = []
    if template.tracks is None:
        subtitles, bubbles = [], []
        for item in source:
            subtitles.append(text("subtitle", item.text, item.start_time, item.end_time))
            if item.keyword:
                bubbles.append(text("bubble", item.keyword,
                                    item.start_time, item.end_time))
        subtitle_tracks = [{"SubtitleTrackClips": subtitles}]
        if request.title and request.title.strip():
            subtitle_tracks.append({"SubtitleTrackClips": [text(
                "title", request.title, source[0].start_time, source[0].end_time,
            )]})
        if bubbles:
            subtitle_tracks.append({"SubtitleTrackClips": bubbles})
    audio_tracks = [{"AudioTrackClips": [{
        "MediaURL": request.audio_url, "In": 0, "Out": duration,
        "TimelineIn": 0, "TimelineOut": duration,
    }]}]
    music = request.pack_rules.background_music
    if music.audio_switch:
        audio_tracks.append({"AudioTrackClips": [{
            "MediaURL": music.audio_url, "In": 0, "TimelineIn": 0, "TimelineOut": duration,
            "LoopMode": True, "Effects": [{"Type": "Volume", "Gain": music.volume}],
        }]})
    timeline = {"VideoTracks": [{"VideoTrackClips": clips}], "AudioTracks": audio_tracks,
                "SubtitleTracks": subtitle_tracks}
    effects = [{"EffectTrackItems": [{"Type": kind, **parameters[key],
                                     "TimelineIn": 0, "TimelineOut": duration}]}
               for key, kind in (("filter", "Filter"), ("vfx", "VFX"))
               if key in parameters]
    if template.tracks is not None:
        for track in template.tracks:
            track_parameters = parameters_for(track.editor)
            applied = resolve_track(track, duration, fps)
            if applied.notice:
                warnings.append(f"对象 {track.id}：{applied.notice}")
            if applied.end <= applied.start:
                continue
            if track.target in ("filter", "vfx"):
                effects.append({"EffectTrackItems": [{
                    "Type": "Filter" if track.target == "filter" else "VFX",
                    **track_parameters[track.target], "TimelineIn": applied.start, "TimelineOut": applied.end,
                }]})
            elif track.target == "transition":
                # 文案合成保持音频时间，转场关联规则指定的切换位置，连续素材在此分为两个片段。
                boundary = applied.end
                for index, clip in enumerate(clips):
                    if clip["TimelineIn"] < boundary < clip["TimelineOut"]:
                        following = {**clip, "Effects": [dict(effect) for effect in clip["Effects"]], "TimelineIn": boundary}
                        if clip["Type"] == "Video":
                            source_boundary = clip["In"] + boundary - clip["TimelineIn"]
                            clip["Out"] = source_boundary
                            following["In"] = source_boundary
                        else:
                            clip["Duration"] = boundary - clip["TimelineIn"]
                            following["Duration"] = following["TimelineOut"] - boundary
                        clip["TimelineOut"] = boundary
                        clips.insert(index + 1, following)
                        break
                index = next((i for i, clip in enumerate(clips[:-1]) if abs(clip["TimelineOut"] - boundary) < 1e-8), None)
                seconds = applied.end - applied.start
                if index is None or min(clips[index]["TimelineOut"] - clips[index]["TimelineIn"], clips[index + 1]["TimelineOut"] - clips[index + 1]["TimelineIn"]) + 1e-8 < seconds:
                    raise ValueError(f"转场对象 {track.id} 的相邻片段不足以容纳持续时间")
                clips[index]["Effects"].append({"Type": "DLTransition", **track_parameters["transition"], "Duration": seconds})
            else:
                # 标题使用请求文字；字幕和关键词只在对象的有效区间内显示，保持业务文字来源。
                contents = [(request.title, applied.start, applied.end)] if track.target == "title" else [
                    (item.text if track.target == "subtitle" else item.keyword,
                     max(item.start_time, applied.start), min(item.end_time, applied.end)) for item in source
                ]
                text_clips = []
                for content, start, end in contents:
                    if not content or not content.strip() or end <= start:
                        continue
                    segment_track = track.model_copy(update={"start_mode": "seconds", "start": start, "duration": end - start})
                    segment = resolve_track(segment_track, duration, fps)
                    if segment.notice:
                        warnings.append(f"对象 {track.id}：{segment.notice}")
                    if segment.end > segment.start:
                        text_clips.append(text(track.target, content, segment.start, segment.end, segment.editor, track_parameters))
                if text_clips:
                    timeline["SubtitleTracks"].append({"SubtitleTrackClips": text_clips})
    if effects:
        timeline["EffectTracks"] = effects
    return timeline, warnings

File: client/src/features/templates/timeline.ts
/** 将模板草稿转换为 SDK 5.2.2 母版、文字和特效轨道，并生成时间轴展示数据。 */
import {
  effectGroups,
  type Draft,
  type EffectAsset,
  type EffectKey,
  type TextRole,
  type MasterVideo,
} from "./model";
import type { TimelineAction, TimelineRow } from "@xzdarcy/timeline-engine";
import { defaultEditor } from "./model";
import { isTextTarget } from "./effects";
import { previewDuration, resolveTrack, trackDraft, trackLabel } from "./tracks";

/** 预览轨道保留源素材范围，缩略图采样使用源时间，游标使用预览时间。 */
export interface PreviewClip extends TimelineAction {
  label: string;
  url: string;
  sourceIn: number;
  sourceOut: number;
  transition?: { start: number; end: number };
  targetId?: string;
}

/** 轨道从已经应用的 SDK Timeline 派生；重叠片段分行展示，保持实际时间范围。 */
export function buildPreviewRows(timeline: ReturnType<typeof buildTimeline>): (Omit<TimelineRow, "actions"> & { actions: PreviewClip[] })[] {
  return [...timeline.VideoTracks.flatMap((track, trackIndex) =>
    track.VideoTrackClips.map((clip, index) => {
      const next = track.VideoTrackClips[index + 1];
      const transition = clip.Effects.some((item) => item.Type === "Transition") && next
        ? { start: Math.max(clip.TimelineIn, next.TimelineIn), end: Math.min(clip.TimelineOut, next.TimelineOut) }
        : undefined;
      return {
        id: `video-${trackIndex}-${index}`,
        actions: [{
          id: `clip-${trackIndex}-${index}`,
          effectId: "video",
          start: clip.TimelineIn,
          end: clip.TimelineOut,
          movable: false,
          flexible: false,
          label: `母版视频片段 ${index + 1}`,
          url: clip.MediaURL,
          sourceIn: clip.In,
          sourceOut: clip.Out,
          transition: transition && transition.end > transition.start ? transition : undefined,
        }],
      };
    }),
  ), ...timeline.previewRows];
}

/** 每秒约取一张源素材缩略图，使用区间中心，避免在片段结束边界采样。 */
export function thumbnailTimes(clip: Pick<PreviewClip, "sourceIn" | "sourceOut">): number[] {
  const { sourceIn, sourceOut } = clip;
  if (!Number.isFinite(sourceIn) || !Number.isFinite(sourceOut) || sourceIn < 0 || sourceOut <= sourceIn)
    throw new Error("视频片段的源时间范围无效");
  const count = Math.min(20, Math.ceil(sourceOut - sourceIn));
  return Array.from({ length: count }, (_, index) => sourceIn + (index + 0.5) * (sourceOut - sourceIn) / count);
}

/** SDK 效果允许混合文字、数值参数；仅从可信目录和经过范围校验的配置生成。 */
type Effect = Record<string, string | number>;

/** 校验数值输入，不让空值、NaN 或越界参数进入 SDK。 */
function number(
  value: number,
  min: number,
  max: number,
  label: string,
  integer = false,
): number {
  if (
    !Number.isFinite(value) ||
    value < min ||
    value > max ||
    (integer && !Number.isInteger(value))
  ) {
    throw new Error(`${label}须为 ${min}～${max}${integer ? " 的整数" : ""}`);
  }
  return value;
}

/** 生成十秒组合预览，转场重叠量使用已保存时长；空文字不会生成轨道。 */
function buildLegacyTimeline(draft: Draft, catalog: EffectAsset[], target?: string) {
  // 空配置回退到原示例；URL 原生解析支持本地路径，避免重复编码已转义的直链。
  const video = new URL(
    import.meta.env.VITE_PREVIEW_VIDEO_URL?.trim() ||
      "https://ice-pub-media.myalicdn.com/vod-demo/最美中国纪录片-智能字幕.mp4",
    window.location.origin,
  ).href;
  const config = draft.editor;
  const byId = new Map(catalog.map((item) => [item.id, item]));
  // 效果类别与字段绑定，拒绝错误类别及未知目录条目。
  const effect = (key: EffectKey): Effect => {
    if (!config[key]) return {};
    const item = byId.get(config[key]);
    if (!item || item.category !== effectGroups[key])
      throw new Error(`效果不在对应目录中：${config[key]}`);
    return { ...item.parameters };
  };
  // SDK 小于 1 才识别为相对坐标，100% 需要限制在边界内。
  const position = (value: number) =>
    Math.min(number(value, 0, 100, "位置") / 100, 0.9999);
  const text = (role: TextRole) => {
    const content = role === "bubble" ? config.bubbleText : config[role];
    const size = number(config[`${role}Size`], 12, 120, "字号", true);
    if (
      config[`${role}Loop`] &&
      (config[`${role}In`] || config[`${role}Out`])
    ) {
      throw new Error("同一类文字的循环动画不能与入场、出场同时使用");
    }
    const motion: Effect = {
      ...effect(`${role}In`),
      ...effect(`${role}Out`),
      ...effect(`${role}Loop`),
    };
    for (const type of ["In", "Out"] as const) {
      const duration = number(
        config[`${role}${type}Duration`],
        1 / 30,
        3,
        "动画时长",
      );
      if (config[`${role}${type}`]) motion[`AaiMotion${type}`] = duration;
    }
    return {
      Type: "Text",
      Content: role === "bubble" ? content : wrapPreviewText(content, size),
      TimelineIn: 0,
      TimelineOut: 10,
      X: position(config[`${role}X`]),
      Y: position(config[`${role}Y`]),
      Alignment: "Center",
      Font: "Alibaba PuHuiTi",
      FontSize: size,
      FontColor: "#FFFFFF",
      Outline: 0,
      ...effect(role === "bubble" ? "bubble" : `${role}Flower`),
      ...motion,
      ...(role === "bubble" ? { Width: 0.5 } : {}),
    };
  };
  const effects: Effect[] = [{ Type: "Volume", Gain: 0 }];
  if (config.filter) effects.push({ Type: "Filter", ...effect("filter") });
  if (config.vfx) effects.push({ Type: "VFX", ...effect("vfx") });
  const overlap = config.transition
    ? number(draft.transition_duration_seconds, 0.1, 3, "转场时长")
    : 0;
  // 仅转场需要两个视频元素；普通预览复用单一素材，降低逐帧纹理上传开销。
  const clip = (start: number, end: number, timelineIn: number) => ({
    Type: "Video",
    MediaURL: video,
    In: start,
    Out: end,
    TimelineIn: timelineIn,
    TimelineOut: timelineIn + end - start,
    Width: 0.9999,
    Height: 0.9999,
    AdaptMode: "Cover",
    Effects: [...effects],
  });
  let clips;
  if (config.transition) {
    // 片段一在第 5 秒后延伸 overlap 秒，片段二从第 5 秒开始，形成真实重叠。
    const first = clip(0, 5 + overlap, 0);
    first.Effects.push({
      Type: "Transition",
      Duration: overlap,
      ...effect("transition"),
    });
    clips = [first, clip(8, 13, 5)];
  } else {
    clips = [clip(0, 10, 0)];
  }
  const texts = [text("title"), text("subtitle")];
  if (config.bubble || target === "bubble") texts.push(text("bubble"));
  return {
    VideoTracks: [{ VideoTrackClips: clips }],
    SubtitleTracks: [
      { SubtitleTrackClips: texts.filter((item) => item.Content.trim()) },
    ],
    AudioTracks: [],
    AspectRatio: "16:9",
    FECanvas: { Width: 800, Height: 450 },
  };
}

/** 母版、文字和全画面效果从同一份实例数据生成，SDK 自动转换后端 Timeline 格式。 */
export function buildTimeline(draft: Draft, catalog: EffectAsset[], media?: MasterVideo) {
  const previewRows: (Omit<TimelineRow, "actions"> & { actions: PreviewClip[] })[] = [];
  const EffectTracks: { EffectTrackItems: Effect[] }[] = [];
  const notices: string[] = [];
  if (!draft.tracks) return { ...buildLegacyTimeline(draft, catalog), EffectTracks, previewRows, notices };
  const duration = previewDuration(draft, media);
  const empty = { ...draft, tracks: undefined, editor: { ...defaultEditor, title: "", subtitle: "", bubbleText: "" } };
  const timeline = buildLegacyTimeline(empty, catalog);
  timeline.SubtitleTracks = [];
  const video = timeline.VideoTracks[0].VideoTrackClips[0];
  if (media) video.MediaURL = media.url;
  video.Out = media?.duration ?? 10;
  video.TimelineOut = duration;
  const seen = new Set<string>();
  let transitionCount = 0;
  for (const source of draft.tracks) {
    const applied = resolveTrack(source, duration);
    if (applied.notice) notices.push(`${trackLabel(source, draft.tracks)}：${applied.notice}`);
    const track = { ...source, ...applied };
    if (!track.id || seen.has(track.id)) throw new Error("特效轨道 ID 必须唯一");
    seen.add(track.id);
    const view = trackDraft(draft, track);
    const resolved = buildLegacyTimeline(view, catalog, track.target);
    if (track.end <= track.start) continue;
    if (isTextTarget(track.target)) {
      const text = resolved.SubtitleTracks[0].SubtitleTrackClips[0];
      if (!text) throw new Error("文字轨道内容不能为空");
      timeline.SubtitleTracks.push({ SubtitleTrackClips: [{ ...text, TimelineIn: track.start, TimelineOut: track.end }] });
    } else if (track.target === "transition") {
      if (++transitionCount > 1) throw new Error("当前母版的两个片段只允许一个转场");
      const effect = resolved.VideoTracks[0].VideoTrackClips[0].Effects.find((item) => item.Type === "Transition");
      if (!effect) throw new Error("转场轨道缺少效果");
      effect.Duration = Math.round((track.end - track.start) * 30) / 30;
      // 源素材首尾相接，播放区间重叠；第二个片段在合成结束时读到源素材末尾。
      timeline.VideoTracks[0].VideoTrackClips = [
        { ...video, Effects: [...video.Effects, effect], Out: track.end, TimelineOut: track.end },
        { ...video, In: track.end, TimelineIn: track.start, Effects: [...video.Effects] },
      ];
    } else {
      const effect = resolved.VideoTracks[0].VideoTrackClips[0].Effects.find((item) => item.Type === (track.target === "filter" ? "Filter" : "VFX"));
      if (!effect) throw new Error("特效轨道缺少效果");
      EffectTracks.push({ EffectTrackItems: [{ ...effect, TimelineIn: track.start, TimelineOut: track.end }] });
    }
    previewRows.push({ id: track.id, actions: [{
      id: track.id, targetId: track.id, effectId: track.target,
      start: track.start, end: track.end, movable: true, flexible: true,
      label: trackLabel(track, draft.tracks), url: "", sourceIn: 0, sourceOut: 0,
    }] });
  }
  if (media) {
    timeline.AspectRatio = `${media.width}:${media.height}`;
    timeline.FECanvas = { Width: 800, Height: Math.round(800 * media.height / media.width) };
  }
  return { ...timeline, EffectTracks, previewRows, notices };
}

/** 按字素折行并保留显式换行；不把 emoji 或组合字符拆开。 */
function wrapPreviewText(content: string, size: number): string {
  const context = document.createElement("canvas").getContext("2d");
  if (!context) return content;
  context.font = `${size}px "ims-preview"`;
  const segmenter = new Intl.Segmenter("zh", { granularity: "grapheme" });
  // ponytail: 测量字体宽度不包含花字外扩；需逐像素一致时改用 SDK 字形测量。
  return content
    .split(/\r\n?|\n/)
    .map((paragraph) => {
      const lines = [""];
      for (const { segment } of segmenter.segment(paragraph)) {
        const last = lines.length - 1;
        if (
          lines[last] &&
          context.measureText(lines[last] + segment).width > 720
        )
          lines.push(segment);
        else lines[last] += segment;
      }
      return lines.join("\n");
    })
    .join("\n");
}
